Add basic ability to download SVGs from D3 panels

huddlej opened the pull request:
Adds download buttons to the line and logo plot panels and links these to a downloadSVG function that serializes the SVG XML of each panel and sends it to the user as a downloaded file. This commit does not add the ability to download the PDB panel. The PDB panel uses Canvas instead of SVG, so we need to use NGL's built-in mechanisms or another approach to support downloads.

Edit: Also adds download button to the protein plot panel linked to the NGL Viewer's `makeImage` function which exports a PNG.

Resolves #147.

huddlej committed:
Add basic ability to download SVGs from D3 panels

Adds download buttons to the line and logo plot panels and links these to a
downloadSVG function that serializes the SVG XML of each panel and sends it to
the user as a downloaded file. This commit does not add the ability to download
the PDB panel. The PDB panel uses Canvas instead of SVG, so we need to use NGL's
built-in mechanisms or another approach to support downloads.

Related to #147.

skhilton commented:
@huddlej: sweet! I'll take a closer look at this in a bit. 

I think we can get the pdb image this way: http://nglviewer.org/ngl/api/class/src/stage/stage.js~Stage.html#instance-method-makeImage

Do you want me to look into that and open a separate pull request or to implement that and push here?

huddlej commented:
Great! I looked into `makeImage` and a) didn't fully understand how to deal with a promise that returns a blob and b) didn't see how to get it to return an SVG or PDF object instead of a PNG.

If you want to add that to this PR, that'd be awesome!

huddlej committed:
Add support for downloading protein panel view as PNG

The protein panel's 3D graphics are rendered with Canvas and WebGL making their
export to SVG technically tricky. The library we use to render the PDB files,
NGL [1], provides a `makeImage` function that exports a PNG. We use that
function here and set the export image quality high enough that it could be used
in a lab notebook or potentially a supplemental figure.

[1] http://nglviewer.org/

huddlej committed:
Vertically align protein and data URL text areas and dropdowns

skhilton commented:
@huddlej: Looks good! I was able to download all three panels and it works when you zoom in on the line plot. The only comment is that the fonts are different in the svgs compared to the browser but I don't think that needs to be fixed (and might be hard!) for this PR to be merged.
